daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/SAXInfosetOutputter.scala

```diff
@@ -0,0 +1,213 @@
+        ch.startPrefixMapping(prefix, ns)
+      }
+
+      if (isNilled(diSimple)) { // TODO split this up into isNilled, hasValue or else (empty)
+        val attrs = new AttributesImpl()
+        attrs.addAttribute(XMLUtils.XSI_NAMESPACE, "nil", "xsi:nil", "", "true")
+        ch.startElement(ns, nsb.local, nsb.toQNameString, attrs)
+      } else if (diSimple.hasValue)  {
+        ch.startElement(ns, nsb.local, nsb.toQNameString, new AttributesImpl())
+        val text =
+          if (diSimple.erd.optPrimType.get.isInstanceOf[NodeInfo.String.Kind]) {
+            // TODO Should we use XMLInfosetOutputter remapped instead?
+            XMLUtils.remapXMLIllegalCharactersToPUA(diSimple.dataValueAsString)
+          } else {
+            diSimple.dataValueAsString
+          }
+        val arr = text.toCharArray
+        ch.characters(arr,0, arr.length)
+      } else { //empty element
+        ch.startElement(ns, nsb.local, nsb.toQNameString, new AttributesImpl())
+      }
+      true
+    } catch {
+      case _: SAXException  => false
+    }
+  }
+
+  /**
+   * Called by Daffodil internals to signify the end of a simple element.
+   *
+   * @param diSimple the simple element that is ended. Various fields of
+   *                 DISimple can be accessed to determine things like the
+   *                 value, nil, name, namespace, etc.
+   * @return true on sucess, false if there was an error and Daffodil should stop all
+   *         future calls to the InfosetOutputter
+   */
+  override def endSimple(diSimple: DISimple): Boolean = {
+    val ns = diSimple.erd.thisElementsNamespace
+    val prefix = diSimple.erd.thisElementsNamespacePrefix
+    val nsb = diSimple.namedQName
+    try {
+      ch.endElement(ns, nsb.local, nsb.toQNameString)
+      // TODO only called endPrefix is there is a prefix
+      ch.endPrefixMapping(prefix)
+      true
+    } catch {
+      case _: SAXException => false
+    }
+  }
+
+  /**
+   * Called by Daffodil internals to signify the beginning of a complex element.
+   *
+   * @param diComplex the complex element that is started. Various fields of
+   *                  DIComplex can be accessed to determine things like the
+   *                  nil, name, namespace, etc.
+   * @return true on sucess, false if there was an error and Daffodil should stop all
+   *         future calls to the InfosetOutputter
+   */
+  override def startComplex(diComplex: DIComplex): Boolean = {
+    val ns = if (diComplex.erd.thisElementsNamespace.isNoNamespace) "" else diComplex.erd.thisElementsNamespace.toString
+    val prefix = if (diComplex.erd.thisElementsNamespacePrefix == null) "" else diComplex.erd.thisElementsNamespacePrefix
+    val nsb = diComplex.namedQName
+    try {
+      if (prefix.nonEmpty && ns.nonEmpty) {
+        ch.startPrefixMapping(prefix, ns)
+      }
+
+      // TODO how about empty elements
+      if (isNilled(diComplex)) {
+        // TODO how to set as nilled in SAX
```

Comment:
I think you're doing this right. As with startSimple, it probably makes sense to split this into separate vals. Maybe even create some a new function to create the nilAttribute or empty AttributesImpl so that startSimple/startComplex can share the logic.